CenterForOpenScience/modular-file-renderer
# [SVCS-571] High Resolution Image with Zoom Feature

## Ticket

~~https://openscience.atlassian.net/browse/SVCS-570~~
https://openscience.atlassian.net/browse/SVCS-571

## Purpose

* Enable high resolution images in MFR

* Enable zoom feature in MFR for images

* Credit goes to @AddisonSchiller Replaces https://github.com/CenterForOpenScience/modular-file-renderer/pull/302.

## Changes

### Bump Up Image Resolution

* MFR used to downsize images to 1200x1200 for render. Bump up the max resolution to 2400 x 2400.

### Enable Zoom

* Allow user to click on the image and view the high resolution image
  * Reference: [jquery.zoom](https://github.com/jackmoore/zoom)

![hi-res-image-after-click](https://user-images.githubusercontent.com/3750414/36313076-7be18f4a-12fe-11e8-89aa-2a1e2538d93e.png)

* Allow user to click and zoom in/out on the image with mouse wheel scroll
  * Reference: [jquery.mousewheel](https://github.com/jquery/jquery-mousewheel)

![further-zoom-using-mouse-wheel](https://user-images.githubusercontent.com/3750414/36313082-82c7506a-12fe-11e8-951d-63dfa69a66f7.png)

* Added a "popover" message as zoom instructions
  * Reference: [jsc3d rendering](https://github.com/CenterForOpenScience/modular-file-renderer/blob/develop/mfr/extensions/jsc3d/templates/viewer.mako#L11)

![no-zoom-with-instructions](https://user-images.githubusercontent.com/3750414/36313031-5f94564c-12fe-11e8-8501-abd97c1c7df2.png)


## Side effects

* Page scroll are disabled when mouse cursor are above the image. Need to move the cursor away or grab the scroll bar to scroll down.

* If an image is eligible for the zoom feature, three images (two in the DOM and one in memory) are loaded. Along with the resolution bump up, this may slow down image rendering for large images (especially when cache is disabled).

## QA Notes

This note is for both QA and CR.

### What to Verify:

- Images are rendered correctly
- Hovering on the question icon pops up the zoom instruction
- Follow the instructions to see if it make sense and if zoom works as expected
- Zoom (and instructions) is enabled only for images that has a height greater than 200
- Monitor how long it takes to render an image (expecting slowdown)

### What to Test

Here is a [public staging OSF project](https://staging.osf.io/s5agd/) that contains all my test files. Please test other images you have as well.

- Images in different browsers (Firefox, Chrome, Safari, IE 11 and Edge)
   - Locally, I only tested Firefox, Chrome and Safari
- Images of different types
   - This shouldn't be of any problem since all images are converted to JPEG for rendering
- Images of different sizes
    - Images greater larger than 2400x2400
    - Images that are downsized by browser to fit in screen
      - Click to view Hi-Res and enable zoom
      - Scroll to zoom in/out base on the Hi-Res version
    - Images that are not downsized
      - Click to enable zoom and scroll to zoom
    - Images that have a height less than 200
      - Zoom is disabled and there is no instructions
    - Images of different shape
      - Images with large height but small width
      - Images with small height but large height
        - Only the top of the image are displayed, user needs to scroll the page down
          - See Side Effects for how
        - Image stay zoomed when user scrolls up and own the page

## Deployment Notes

- [ ] Monitor MFR to see if there is a dramatic performance hit due to:
  - This PR increase the base image size from 1200x1200 to 2400x2400
  - The zoom feature loads three copy of the image on the page
